### PR TITLE
Reworking SDK

### PR DESCRIPTION
**Pull request by DavidAntaramian**
:warning: Needs some good review and feedback

@joshdholtz You can take a look at this. I haven't added in much testing yet because I really need to test the inverse of what Nocilla offers (i.e., it doesn't so much matter what's coming in as what's going out except in the case of the reinforce call which has a test).

### CODE REVIEW

File: Dopamine.podspec [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  s.summary      = "A Swift client for http://usedopamine.com/"
+
+  s.description  = <<-DESC
+                    Dopamine eases connection to the Dopamine API
+                    which helps you make your user's actions habitual.
+                   DESC
+
+  s.homepage     = "http://usedopamine.com/"
+  s.license      = { :type => "MIT", :file => "LICENSE.MD" }
+  s.author             = { "Ramsay Brown" => "ramsay@usedopamine.com" }
```
Comment (joshdholtz) [inline Dopamine.podspec:13]: @DavidAntaramian Add yourself into this :grin:

File: Dopamine/Dopamine.swift [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+import SwiftyJSON
+
+/// Generic settings for connecting to Dopamine
+typealias Settings = (appID: String, apiKey: String, token: String, versionID: String, pairings: ActionPairings)
+/// Reinforcers are rewards or feedback for the end user which are primarily identified by name
+typealias Reinforcer = (name: String, type: FunctionType, constraint: [String], objective: [String] )
+/// Actions have a 1:n pairing with possible rewards and feedback (Reinforcer)
+typealias ActionPairing = (name: String, reinforcers: [Reinforcer])
+/// A map of names to action pairings
+typealias PairingDictionary = [String:ActionPairing]
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Swift style would be to make this `[String: ActionPairing]` (with a space in between)

File: Dopamine/Dopamine.swift [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+/// Convenience type for the request payload
+typealias APIPayload = [String:AnyObject]
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Also here `[String: AnyObject]`

File: Dopamine/Dopamine.swift [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+/// Closure of the form {(statusCode, functionName) -> Void} that is provided a callback on reinforcements
+public typealias ReinforcementCallback = ((Int, String) -> Void)
+
+/**
+    The API is namespaced by suffixes. This enum defines the appropriate sufixes to be added to the
+    end of a URL
+*/
+enum APIEndpoint : String {
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Swift style would be to put the colon right after the enum name `enum APIEndpoint: String {`

File: Dopamine/Dopamine.swift [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    case Init = "init"
+    case Reinforce = "reinforce"
+    case Track = "track"
+}
+
+/**
+    The API can handle reward and feedback functions for each action pairing
+*/
+enum FunctionType : String {
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Same `enum FunctionType: String {`

File: Dopamine/Dopamine.swift [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+/**
+    Maintains pairings of action names to the relevant reward and feedback identifies
+*/
+public class ActionPairings {
+    var store = PairingDictionary()
+    
+    /// All the feedback functions that have been registered so far
+    var feedbackFunctions : Set<String> {
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Remove space before the colon here too

File: Dopamine/Dopamine.swift [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        get {
+            var bid = ""
+            for (name, pairing) in store {
+                bid += pairing.name
+                bid += reduce(pairing.reinforcers,"",{$0 + $1.name})
+            }
+            if let s = bid.sha1() {
+                return s
+            }
+            return ""
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift]: @DavidAntaramian Could shorten this to `return bid.sha1() ?? ""`

File: Dopamine/Dopamine.swift [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+        let standardizedPayload = basePayload.merge(t)
+        let requestPayload = standardizedPayload.merge(payload)
+        let url = buildEndpointURL(endpoint)
+        Alamofire.request(.POST, url, parameters: requestPayload, encoding: .JSON)
+                 .responseJSON { _, netResponse, JSON, _ in
+                    if endpoint == .Reinforce {
+                        if let r = netResponse, j: AnyObject = JSON, functionName = j["reinforcementFunction"] as? String, c = callback {
+                            c(r.statusCode, functionName)
+                        }
+                    }
```
Comment (joshdholtz) [inline Dopamine/Dopamine.swift:256]: @DavidAntaramian You _may_ want to have a callback that happens when an error occurs (just so that the UI isn't waiting for response on something for forever)

**Comment by joshdholtz** [comment]
@DavidAntaramian This looks super good :raised_hands: The biggest thing is maybe calling the closure when an error in the API happens and also some Swift style things.

Back to you :cake:

**Comment by DavidAntaramian** [comment]
@joshdholtz Changes made ^^ I tried using a case statement, not sure if I did it correctly though

**Comment by joshdholtz** [comment]
@DavidAntaramian Everything looks good to me :blush: :+1:
